Add option to ignore function coverage

Spacetown opened the pull request:
If coverage isn't ignored, the JSON output is extended with property 'gcovr/function' with value True/False.

close #398

latk commented on gcovr/json_generator.py:
```diff
@@ -172,6 +172,7 @@ def _json_from_line(line):
     json_line['count'] = line.count
     json_line['line_number'] = line.lineno
     json_line['gcovr/noncode'] = line.noncode
+    json_line['gcovr/function'] = line.function
```
If adding a new JSON field we should also update deserialization, and possibly bump the version number. In the interest of compression we might also skip this field unless it is set?

But as discussed above, we might not need this field in the first place.

Spacetown replied:
I'll remove this field and set always noncode for the function line as mentioned above.

Spacetown commented on gcovr/configuration.py:
```diff
@@ -884,6 +884,13 @@ def choose(_cls, choices, default=None):
              "Default: {default!s}.",
         action="store_true",
     ),
+    GcovrConfigOption(
+        "exclude_function_coverage", ['--exclude-function-coverage'],
+        group="gcov_options",
+        help="Exclude coverage from lines defining a function "
+             "Default: {default!s}.",
+        action="store_true",
+    ),
```
I agree with you.

Spacetown commented on gcovr/gcov.py:
```diff
@@ -298,25 +300,40 @@ def parse_line(
+                    return True
+                self.coverage.line(self.lineno).function = True
             if self.excluding or is_non_code(code):
                 self.coverage.line(self.lineno).noncode = True
```
I'Ve tried this, but also the counter and and noncode has to be set outside because if the line is excluded it must not be added in the if/elif construct. I set the noncode also if the function coverage isn't excluuded.
I'm not sure if this is correct.

latk replied:
Ok, that looks better but it the code seems incorrect – all tests were updated, which can't be the case (?). Apparently now all function lines are being excluded by default?

I think the problem is this snippet:

```python
if not (last_was_function_marker and exclude_function_lines):
    if last_was_function_marker:
        noncode = True
```

We now have `last_was_function_marker = True` and `exclude_function_lines = False`. The outer condition will evaluate to True, but the inner condition will set `noncode = True` nevertheless.

I think that section should be rewritten like this:

```python
if exclude_function_coverage and last_was_function_marker:
    noncode = True

if noncode:
    self.coverage.line(self.lineno).noncode = True
if count is not None:
    self.coverage.line(self.lineno).count += count
```

Or maybe this `noncode = True` line should be a `return True`?

Spacetown replied:
That's why I wasn't sure. If we set the line flag `noncode = True` we shouldn't increment the line counter. In my opionion a noncode line can't have a counter not zero. I'll check your suggestions, this looks better. I'll think it should be:
```python
if noncode:
    self.coverage.line(self.lineno).noncode = True
elif count is not None:
    self.coverage.line(self.lineno).count += count```

latk reviewed (approved):
Missing a changelog entry, otherwise ready to go :) Thanks for working on this.

Potential future directions:

* The gcov parser state machine is growing more and more complex. Perhaps it would be possible to separate this into a textual part that generates events for each line, and then various event consumers with simpler state machines that update the coverage data structure. For example when reviewing this code, I was unsure where the `last_was_function_marker` was being reset because it was spread over so many lines.

* For each compiler version, we are generating all output formats. This requires extra effort when updating reference files and slows down the test suite. It would be sufficient to only test JSON output for individual compilers, since the other outputs can be generated from JSON without depending on any compiler. So we could move from O(compilers × outputs) tests to O(compilers + outputs). In some places, the test suite has already been rewritten to use JSON as an intermediate format but the JSON isn't reused between tests and the tests still depend on freshly generated JSON (requiring a compiler).

Spacetown commented:
@latk I've rebased and added the issue to the changelog.
The event based parser is a good idea but first I'll create an abstract interface for readers and writers as discussed in #443.
I think we can't remove the call of gcc in the whole test suite because also the gcov parser must be tested. But the tests of the writer can be improved here.

latk commented on CHANGELOG.rst:
```diff
@@ -33,6 +33,7 @@ Improvements and new features:
  - Only remove :option:`--root` path at the start of file paths. (:issue:`452`)
  - Fix coverage report for cmake ninja builds with given in-source object-directory. (:issue:`453`)
  - Add issue templates. (:issue:`461`)
+ - Add :option:`--exclude-function-coverage` to exclude the line of the function definition in the coverage report. (:issue:`430`)  
```
Building the sphinx docs found an error for us:

```suggestion
 - Add :option:`--exclude-function-lines` to exclude the line of the function definition in the coverage report. (:issue:`430`)  
```

Spacetown replied:
Yes, currently I'm rebasing and updating the test data.
